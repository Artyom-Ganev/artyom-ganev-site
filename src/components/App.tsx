import React from "react";
import { Route, Switch } from "react-router-dom";
import "./App.scss";
import Menu from "./Navigation/Menu";
import Main from "./Page/Main";
import About from "./Page/About";
import Blog from "./Page/Blog";
import Career from "./Page/Career";
import Contacts from "./Page/Contacts";
import NotFound from "./Page/NotFound";
import IState from "../interface/IState";
import { connect, DispatchProp } from "react-redux";
import { MAIN } from "../constants/Pages";

export interface IProperties extends DispatchProp {
  navigation: { activePage: string };
}

/**
 * Main application component
 */
const App = ({ dispatch, navigation }: IProperties) => (
  <div className="flexBox flexColumn">
    <Menu dispatch={dispatch} activePage={navigation.activePage} />
    <div className="components-app__background" />
    <div className="flexBox flexColumn components-app__content">
      <Switch>
        <Route exact={true} path="" component={Main} />
        <Route path="about" component={About} />
        <Route path="blog" component={Blog} />
        <Route path="career" component={Career} />
        <Route path="contacts" component={Contacts} />
        <Route component={NotFound} />
      </Switch>
    </div>
    <div className="components-app__rights">Â© 2019 All rights reserved</div>
  </div>
);

/**
 * Component state to props mapping function
 * @param {IState} state
 * @param {Object} ownProps
 */
const mapStateToProps = (
  { navigation }: IState,
  ownProps: { location: { pathname: string } }
) => {
  const { activePage } = navigation;
  if (activePage) {
    return { navigation: { activePage } };
  }
  // We need to set default page on first login
  const paths = ownProps.location.pathname.split("/");
  return { navigation: { activePage: paths[1] || MAIN.id } };
};

export default connect(mapStateToProps)(App);
