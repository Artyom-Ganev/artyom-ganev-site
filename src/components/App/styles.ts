import { createStyles, makeStyles } from '@material-ui/core/styles';
import { rybinsk } from '../../assets';

const APP_HEADER_HEIGHT = '37px';
const APP_FOOTER_HEIGHT = '20px';

export default makeStyles(() =>
  createStyles({
    app: {
      backgroundImage: `url('${rybinsk}')`,
      backgroundSize: 'cover',
      bottom: 0,
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,
    },
    appContainer: {
      backgroundColor: 'white',
      height: '100vh',
      opacity: '0.7',
      '&:hover': {
        opacity: '0.8',
      },
    },
    appContent: {
      bottom: APP_FOOTER_HEIGHT,
      position: 'absolute',
      top: APP_HEADER_HEIGHT,
    },
    appFooter: {
      bottom: 0,
      height: APP_FOOTER_HEIGHT,
      left: 0,
      position: 'fixed',
      right: 0,
      textAlign: 'center',
    },
    appHeader: {
      height: APP_HEADER_HEIGHT,
    },
  })
);
