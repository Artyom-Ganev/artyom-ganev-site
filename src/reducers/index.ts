import { combineReducers } from 'redux';
import { handleActions } from 'redux-actions';
import * as actions from 'src/actions';
import { IBlog, IBlogProps } from 'src/interface/IBlog';
import { ICareer, ICareerProps } from 'src/interface/ICareer';
import { IAppState } from 'src/interface/IState';

// fix redux-actions TypeScript bad support
const appLoading = `${actions.appLoading}`;
const blogLoadList = `${actions.blogLoadList}`;
const careerLoadList = `${actions.careerLoadList}`;

/**
 * App state reducer
 */
export const app = handleActions<IAppState>(
  {
    [appLoading](state: IAppState, { payload: { loading } }: { payload: { loading: boolean } }) {
      return { ...state, loading };
    },
    [blogLoadList](state: IAppState) {
      return { ...state, loading: false };
    },
    [careerLoadList](state: IAppState) {
      return { ...state, loading: false };
    },
  },
  { loading: false }
);

/**
 * Blog actions reducer
 */
export const blog = handleActions(
  {
    [blogLoadList](state: IBlogProps, { payload: { items } }: { payload: { items: IBlog[] } }) {
      const newItems = state.items;
      items.forEach((item) => {
        newItems[item.id] = item;
      });
      return { ...state, items: newItems };
    },
  },
  { items: {}, detail: {} }
);

/**
 * Career actions reducer
 */
export const career = handleActions(
  {
    [careerLoadList](state: ICareerProps, { payload: { items } }: { payload: { items: ICareer[] } }) {
      const newItems = state.items;
      items.forEach((item) => {
        newItems[item.id] = item;
      });
      return { ...state, items: newItems };
    },
  },
  { items: {}, detail: {} }
);

export default combineReducers({ app, blog, career });
