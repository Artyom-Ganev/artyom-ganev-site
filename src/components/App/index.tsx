import { createBrowserHistory } from 'history';
import React, { FunctionComponent, useEffect, useState } from 'react';
import { Router } from 'react-router-dom';
import { Content, Footer, Header } from 'src/components/Layout';

/**
 * Check if the page is narrow
 */
const isNarrow = (): boolean => window.innerWidth <= 800;

/**
 * Resize hook
 */
const useResize = (): boolean => {
  const [narrow, setNarrow] = useState<boolean>(isNarrow());
  useEffect(() => {
    const onResize = (): void => setNarrow(isNarrow());
    window.addEventListener('resize', onResize);
    return (): void => window.removeEventListener('resize', onResize);
  });
  return narrow;
};

export const NarrowContext = React.createContext(isNarrow());

export const history = createBrowserHistory();

/**
 * Main application component
 */
const App: FunctionComponent = () => (
  <Router history={history}>
    <div className='flexBox flexColumn'>
      <Header />
      <NarrowContext.Provider value={useResize()}>
        <Content />
      </NarrowContext.Provider>
      <Footer />
    </div>
  </Router>
);

export default App;
