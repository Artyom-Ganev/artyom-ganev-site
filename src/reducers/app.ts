import {
  APP_LOADING,
  BLOG_LOAD_LIST,
  BLOG_LOAD_LIST_ERROR,
  BLOG_LOAD_LIST_SUCCESS,
  CAREER_LOAD_LIST,
  CAREER_LOAD_LIST_ERROR,
  CAREER_LOAD_LIST_SUCCESS,
} from 'src/actions';
import { AppActionTypes, IAppState } from 'src/types';

const initialState: IAppState = {
  loading: false,
};

/**
 * App state reducer
 */
const app = (state: IAppState = initialState, action: AppActionTypes): IAppState => {
  switch (action.type) {
    case APP_LOADING: {
      const { loading } = action;
      return { ...state, loading };
    }
    case BLOG_LOAD_LIST: {
      return { ...state, loading: true };
    }
    case BLOG_LOAD_LIST_SUCCESS: {
      return { ...state, loading: false };
    }
    case BLOG_LOAD_LIST_ERROR: {
      return { ...state, loading: false };
    }
    case CAREER_LOAD_LIST: {
      return { ...state, loading: true };
    }
    case CAREER_LOAD_LIST_SUCCESS: {
      return { ...state, loading: false };
    }
    case CAREER_LOAD_LIST_ERROR: {
      return { ...state, loading: false };
    }
    default: {
      return state;
    }
  }
};

export default app;
