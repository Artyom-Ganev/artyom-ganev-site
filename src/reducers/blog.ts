import {
  BLOG_LOAD_LIST_ERROR,
  BLOG_LOAD_LIST_SUCCESS,
  BlogActionsType,
  IErrorAction,
  ILoadBlogListSuccessAction,
} from 'src/actions';
import { IBlogState } from 'src/types';

export const initialState: IBlogState = {
  detail: {},
  error: '',
  items: [],
};

/**
 * Blog actions reducer
 */
const blog = (state: IBlogState = initialState, action: BlogActionsType): IBlogState => {
  switch (action.type) {
    case BLOG_LOAD_LIST_SUCCESS: {
      const { items } = action as ILoadBlogListSuccessAction;
      return { ...state, items };
    }
    case BLOG_LOAD_LIST_ERROR: {
      const { error } = action as IErrorAction;
      return { ...state, error };
    }
    default: {
      return state;
    }
  }
};

export default blog;
