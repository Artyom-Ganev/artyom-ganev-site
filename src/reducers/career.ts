import { CAREER_LOAD_LIST_ERROR, CAREER_LOAD_LIST_SUCCESS } from 'src/actions';
import { CareerActionsType, ICareerState, ILoadCareerListSuccessAction } from 'src/types';

const initialState: ICareerState = {
  items: [],
  detail: {},
  error: '',
};

const career = (state: ICareerState = initialState, action: CareerActionsType): ICareerState => {
  switch (action.type) {
    case CAREER_LOAD_LIST_SUCCESS: {
      const { items } = action as ILoadCareerListSuccessAction;
      return { ...state, items };
    }
    case CAREER_LOAD_LIST_ERROR: {
      const { error } = action;
      return { ...state, error };
    }
    default: {
      return state;
    }
  }
};

export default career;
