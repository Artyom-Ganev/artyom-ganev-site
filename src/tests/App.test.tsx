import React from "react";
import renderer from "react-test-renderer";
import App from "../components/App";
import { HashRouter, Route } from "react-router-dom";
import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import { navigation } from "../reducers";

test("App", () => {
  const store = createStore(combineReducers({ navigation }));

  const component = renderer.create(
    <Provider store={store}>
      <HashRouter>
        <Route component={App} />
      </HashRouter>
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
