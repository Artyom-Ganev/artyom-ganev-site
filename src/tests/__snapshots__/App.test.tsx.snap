// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 1`] = `
<div
  className="flexBox flexColumn"
>
  <header
    className="flexBox"
  >
    <nav
      className="flexBox nav-menu__root"
      id="menu"
    >
      <a
        aria-current="page"
        className="nav-menu__item active"
        href="#/"
        onClick={[Function]}
        style={Object {}}
      >
        Main
      </a>
      <a
        aria-current={null}
        className="nav-menu__item"
        href="#/about"
        onClick={[Function]}
      >
        About
      </a>
      <a
        aria-current={null}
        className="nav-menu__item"
        href="#/blog"
        onClick={[Function]}
      >
        Blog
      </a>
      <a
        aria-current={null}
        className="nav-menu__item"
        href="#/career"
        onClick={[Function]}
      >
        Career
      </a>
      <a
        aria-current={null}
        className="nav-menu__item"
        href="#/contacts"
        onClick={[Function]}
      >
        Contacts
      </a>
    </nav>
  </header>
  <div
    className="components-content__background"
  />
  <div
    className="flexBox flexColumn components-content__wrapper"
  >
    <div
      className="flexBox flexColumn page-container__root"
    >
      <div
        className="page-main__root"
      >
        <img
          alt="Artyom"
          className="page-main__image"
          src="artyom.jpg"
          title="artyom.jpg"
        />
        <h2
          className="page-main__contentBlock"
        >
          Hi! My name is Artyom.
        </h2>
        <div
          className="page-main__contentBlock"
        >
          I'm a programmer, based in 
          <a
            href="https://en.wikipedia.org/wiki/Rybinsk"
            rel="noopener noreferrer"
            target="_blank"
          >
            Rybinsk
          </a>
          , Russia. There are some facts about me on this site.
        </div>
        <div
          className="page-main__contentBlock"
        >
          You can contact me if you have some ideas for it's improvement or if you find some bugs, typos etc. I'm still working on it.
        </div>
        <div
          className="page-main__contentBlock"
        >
          Frontend of this site is based on React, Redux, TypeScript and Sass, the sources of this part are located on GitHub 
          <a
            href="https://github.com/Artyom-Ganev/artyom-ganev-site"
            rel="noopener noreferrer"
            target="_blank"
          >
            here
          </a>
          .
        </div>
        <div
          className="page-main__contentBlock"
        >
          Backend uses Spring Boot, Maven, Heroku and PostgreSQL and is located 
          <a
            href="https://github.com/Artyom-Ganev/artyom-ganev-server"
            rel="noopener noreferrer"
            target="_blank"
          >
            here
          </a>
          .
        </div>
      </div>
      <div
        className="flexBox justifyContentCenter page-main__githubContainer"
      >
        <div
          className="page-main__githubContributions"
        />
        <div
          className="flexBox flexGrow-1 flexShrink-1 justifyContentCenter"
        >
          <div
            className="nav-loading__indicator"
          />
        </div>
      </div>
    </div>
  </div>
  <footer
    className="components-footer__container"
  >
    <span>
      © 2019 All rights reserved
    </span>
  </footer>
</div>
`;
